File: algo/lc.141.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (nullptr == head) return head;
        
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast != nullptr && slow != nullptr) {
            slow = slow->next;
            fast = (fast==nullptr) ? nullptr : fast->next;
            fast = (fast==nullptr) ? nullptr : fast->next;
            if (fast != nullptr && fast == slow) {
                return true;
            }
        }
        return false;
    }
};

File: algo/lc.141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head
        while True:
            fast = fast.next if fast is not None else None
            if fast is None or slow is None:
                return False
            elif fast == slow:
                return True
            else:
                fast = fast.next
                slow = slow.next
